pkg/frontend/adminactions/upgrade.go

```diff
@@ -0,0 +1,103 @@
+package adminactions
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/status"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+func (a *adminactions) Upgrade(ctx context.Context) error {
+	err := preUpgradeChecks(ctx, a.oc, a.vNetClient)
+	if err != nil {
+		return err
+	}
+
+	return upgrade(ctx, a.log, a.configClient)
+}
+
+func upgrade(ctx context.Context, log *logrus.Entry, configClient configclient.Interface) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cv, err := configClient.ConfigV1().ClusterVersions().Get("version", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// We don't upgrade if cluster upgrade is not finished
+		if !status.ClusterVersionOperatorIsHealthy(cv.Status) {
+			return fmt.Errorf("not upgrading: previous upgrade in-progress")
+		}
+
+		desired, err := version.ParseVersion(cv.Status.Desired.Version)
+		if err != nil {
+			return err
+		}
+
+		// Get Cluster upgrade version based on desired version
+		stream, err := version.GetUpgradeStream(desired)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("initiating cluster upgrade, target version %s", stream.Version.String())
+
+		cv.Spec.Channel = stream.Channel
+		cv.Spec.DesiredUpdate = &configv1.Update{
+			Version: stream.Version.String(),
+			Image:   stream.PullSpec,
+		}
+
+		_, err = configClient.ConfigV1().ClusterVersions().Update(cv)
+		return err
+	})
+}
+
+func preUpgradeChecks(ctx context.Context, oc *api.OpenShiftCluster, vnet network.VirtualNetworksClient) error {
+	return checkCustomDNS(ctx, oc, vnet)
```

Comment:
Are you anticipating needing to add more to this function? If not, I would just call `checkCustomDNS` directly from `Upgrade`.

Reply:
I think there might be more of these... Hence new function.